src/chart_types/partition_chart/layout/viewmodel/fill_text_layout.ts

```diff
@@ -299,189 +307,254 @@ type GetShapeRowGeometry<C> = (
+    measure: TextMeasure,
+    rawTextGetter: RawTextGetter,
+    valueGetter: ValueGetterFunction,
+    formatter: ValueFormatter,
+    leftAlign: boolean,
+    middleAlign: boolean,
+  ) {
+    const { maxRowCount, fillLabel, fontFamily: configFontFamily } = config;
+    return (allFontSizes: Pixels[][], textFillOrigin: PointTuple, node: QuadViewModel): RowSet => {
+      const container = shapeConstructor(node);
+      const rotation = getRotation(node);
+
+      const layer = layers[node.depth - 1] || {};
+      const verticalAlignment = middleAlign
+        ? VerticalAlignments.middle
+        : node.depth < layers.length
+        ? VerticalAlignments.bottom
+        : VerticalAlignments.top;
+      const fontSizes = allFontSizes[Math.min(node.depth, allFontSizes.length) - 1];
+      const {
+        textColor,
+        textInvertible,
+        fontStyle,
+        fontVariant,
+        fontFamily,
+        fontWeight,
+        valueFormatter,
+        padding,
+      } = Object.assign(
+        { fontFamily: configFontFamily, fontWeight: 'normal', padding: 2 },
+        fillLabel,
+        { valueFormatter: formatter },
+        layer.fillLabel,
+        layer.shape,
+      );
 
-    const valueFont = Object.assign(
-      { fontFamily: config.fontFamily, fontWeight: 'normal' },
-      config.fillLabel && config.fillLabel.valueFont,
-      fillLabel,
-      fillLabel.valueFont,
-      layer.fillLabel,
-      layer.fillLabel && layer.fillLabel.valueFont,
-    );
+      const fillTextColor = getFillTextColor(node.fillColor, textColor, textInvertible);
 
-    const specifiedTextColorIsDark = colorIsDark(textColor);
-    const shapeFillColor = node.fillColor;
-    const { r: tr, g: tg, b: tb, opacity: to } = stringToRGB(textColor);
-    let fontSizeIndex = fontSizes.length - 1;
-    const sizeInvariantFont: Font = {
-      fontStyle,
-      fontVariant,
-      fontWeight,
-      fontFamily,
-    };
-    const allBoxes = getAllBoxes(rawTextGetter, valueGetter, valueFormatter, sizeInvariantFont, valueFont, node);
-    let rowSet = identityRowSet();
-    let completed = false;
-    const rotation = getRotation(node);
-    const container = shapeConstructor(node);
-    const [cx, cy] = textFillOrigins[index];
-
-    while (!completed && fontSizeIndex >= 0) {
-      const fontSize = fontSizes[fontSizeIndex];
-      const wordSpacing = getWordSpacing(fontSize);
-
-      // model text pieces, obtaining their width at the current font size
-      const measurements = measure(fontSize, allBoxes);
-      const allMeasuredBoxes: RowBox[] = measurements.map(
-        ({ width, emHeightDescent, emHeightAscent }: TextMetrics, i: number) => ({
-          width,
-          verticalOffset: -(emHeightDescent + emHeightAscent) / 2, // meaning, `middle`,
-          wordBeginning: NaN,
-          ...allBoxes[i],
-          fontSize, // iterated fontSize overrides a possible more global fontSize
-        }),
+      const valueFont = Object.assign(
```

Comment:
can you add a type here that this merge is going to resolve to?

```suggestion
      const valueFont: MyType = Object.assign(
```

Reply:
See prev. answer